### ISSUE TITLE
Loading external StyledString package removes markdown highlighting

### ISSUE THREAD

**Issue by KristofferC** [JuliaLang/StyledStrings.jl#91]
![image](https://github.com/user-attachments/assets/c7479dbc-183e-4db2-9ea9-2259b14f33c8)

This is when having this package as the active project and loading it.

**Comment by topolarity** [comment]
Do we have a root cause for this yet?

**Comment by tecosaur** [comment]
I don't, but I have noticed that with StyledStrings as the active project you can see this in a Julia REPL session:

```julia
help?> thing
# Observe working syntax highlighting

julia> using StyledStrings

help?> thing
# Syntax highlighting breaks

julia> using Markdown

help?> thing
# Syntax highlighting works again
```

Checking `Base.loaded_modules_order` it seems like two versions of `StyledStrings`, `JuliaSyntaxHighlighting`, and `Markdown` are loaded by doing this:

```julia
julia> Base.loaded_modules_order
22-element Vector{Module}:
 Core
 Base
 Main
 FileWatching
 Libdl
 Artifacts
 SHA
 Sockets
 LinearAlgebra
 OpenBLAS_jll
 libblastrampoline_jll
 Random
 Base64
 StyledStrings
 JuliaSyntaxHighlighting
 Markdown
 InteractiveUtils
 Unicode
 REPL
 StyledStrings
 JuliaSyntaxHighlighting
 Markdown
 
julia> length(unique(Base.loaded_modules_order))
22
```

**Comment by topolarity** [comment]
If I had to guess, StyledStrings has some code that is embedding Face types: https://github.com/JuliaLang/StyledStrings.jl/blob/da41b6a8ea4d4e355159e9c791111c445cc6805c/src/styledmarkup.jl#L725

but then `getface` here is effectively doing a `isa Face` check: https://github.com/JuliaLang/StyledStrings.jl/blob/da41b6a8ea4d4e355159e9c791111c445cc6805c/src/io.jl#L237

which runs into the exact issue I accidentally created with TOML recently (https://github.com/JuliaLang/Pkg.jl/issues/4017#issuecomment-2377589989)

Basically `require_stdlib` means that multiple StyledStrings can exist, which means that the `_ansi_writer` code has to support the `Face` type for _all_ of the loaded StyledStrings.

**Comment by topolarity** [comment]
I think I found the issue - The second copy of StyledStrings contains a default "FACES" dictionary (w/o the Markdown faces) and then overrides Base.write(::IO, ::Base.AnnotatedString) so that any old AnnotatedStrings floating about start seeing the "FACES" only from the second loaded copy of StyledStrings. In particular, the AnnotatedString's created by Markdown are suddenly missing their Face definitions.

Mis-behavior like this is really easy to get with `require_stdlib` + type-piracy, so I've started to document the rules of the game in https://github.com/JuliaLang/julia/pull/56005

The part that matters is:
>  A stdlib must not access any global state that may differ between stdlib copies in type-pirated methods

Given that guideline, I think we only have a few options here:
 - Eliminate the type piracy: Make `AnnotatedString` a different type for each of the StyledStrings copies (move it out of Base)
 - Avoid `FACES` duplication: StyledStrings needs to move into the sysimage, or `require_stdlib` needs to be banned
 - Include a unique, type-based identifier for Annotations so that StyledStrings can resolve `FACES` in a non-type-pirated method
 - Maintain a reference to the relevant `FACES` dict in the AnnotatedString

The basic problem is that a type-pirated method must be simultaneously correct for all copies of a stdlib.

**Comment by christiangnrd** [comment]
This is manifesting as an error when trying to autocomplete with a devved version of StyledStrings.

Typing the `(` is what triggers the error.

```julia-repl
julia> VERSION
v"1.14.0-DEV.3134"

julia> using StyledStrings # In the StyledStrings dev environment

julia> AnnotatedIOBuffer┌ Error: Error in the keymap
│   exception =
│    MethodError: no method matching _mergedface(::StyledStrings.Face)
│    The definition in StyledStrings may have intended to extend (StyledStrings)._mergedface
│    The function `_mergedface` exists, but no method is defined for this combination of argument types.
│    Hint: You may have intended `StyledStrings.Face` rather than `StyledStrings.Face`.
│    
│    Closest candidates are:
│      _mergedface(::StyledStrings.Face)
│       @ StyledStrings ~/.julia/dev/StyledStrings/src/theme.jl:239
│      _mergedface(::Symbol)
│       @ StyledStrings ~/.julia/dev/StyledStrings/src/theme.jl:240
│      _mergedface(::Vector)
│       @ StyledStrings ~/.julia/dev/StyledStrings/src/theme.jl:241
│    
│    Stacktrace:
│      [1] (::Base.MappingRF{typeof(StyledStrings._mergedface), Base.BottomRF{typeof(merge)}})(acc::StyledStrings.Face, x::Any)
│        @ Base reduce.jl:98 [inlined]
│      [2] (::Base.MappingRF{StyledStrings.var"#getface##0#getface##1", Base.MappingRF{typeof(StyledStrings._mergedface), Base.BottomRF{typeof(merge)}}})(acc::StyledStrings.Face, x::@NamedTuple{label::Symbol, value})
│        @ Base reduce.jl:98 [inlined]
│      [3] (::Base.FilteringRF{StyledStrings.var"#getface##2#getface##3", Base.MappingRF{StyledStrings.var"#getface##0#getface##1", Base.MappingRF{typeof(StyledStrings._mergedface), Base.BottomRF{typeof(merge)}}}})(acc::StyledStrings.Face, x::@NamedTuple{label::Symbol, value})
│        @ Base reduce.jl:110 [inlined]
│      [4] _foldl_impl(op::Base.FilteringRF{StyledStrings.var"#getface##2#getface##3", Base.MappingRF{StyledStrings.var"#getface##0#getface##1", Base.MappingRF{typeof(StyledStrings._mergedface), Base.BottomRF{typeof(merge)}}}}, init::Base._InitialValue, itr::Vector{@NamedTuple{label::Symbol, value}})
│        @ Base reduce.jl:60
│      [5] foldl_impl(op::Base.FilteringRF{StyledStrings.var"#getface##2#getface##3", Base.MappingRF{StyledStrings.var"#getface##0#getface##1", Base.MappingRF{typeof(StyledStrings._mergedface), Base.BottomRF{typeof(merge)}}}}, nt::Base._InitialValue, itr::Vector{@NamedTuple{label::Symbol, value}})
│        @ Base reduce.jl:46 [inlined]
│      [6] mapfoldl_impl(f::typeof(StyledStrings._mergedface), op::typeof(merge), nt::Base._InitialValue, itr::Base.Generator{Base.Iterators.Filter{StyledStrings.var"#getface##2#getface##3", Vector{@NamedTuple{label::Symbol, value}}}, StyledStrings.var"#getface##0#getface##1"})
│        @ Base reduce.jl:42 [inlined]
│      [7] mapfoldl(f::typeof(StyledStrings._mergedface), op::typeof(merge), itr::Base.Generator{Base.Iterators.Filter{StyledStrings.var"#getface##2#getface##3", Vector{@NamedTuple{label::Symbol, value}}}, StyledStrings.var"#getface##0#getface##1"}; init::Base._InitialValue)
│        @ Base reduce.jl:173 [inlined]
│      [8] getface(faces::Base.Generator{Base.Iterators.Filter{StyledStrings.var"#getface##2#getface##3", Vector{@NamedTuple{label::Symbol, value}}}, StyledStrings.var"#getface##0#getface##1"})
│        @ StyledStrings ~/.julia/dev/StyledStrings/src/theme.jl:251
│      [9] getface(annotations::Vector{@NamedTuple{label::Symbol, value}})
│        @ StyledStrings ~/.julia/dev/StyledStrings/src/theme.jl:265 [inlined]
│     [10] _ansi_writer(string_writer::typeof(write), io::IOContext{Base.Terminals.TerminalBuffer}, s::SubString{AnnotatedString{String}})
│        @ StyledStrings ~/.julia/dev/StyledStrings/src/io.jl:262
│     [11] ansi_write(f::typeof(write), io::IOContext{Base.Terminals.TerminalBuffer}, s::SubString{AnnotatedString{String}})
│        @ StyledStrings ~/.julia/dev/StyledStrings/src/io.jl:285
│     [12] ansi_write_(f::typeof(write), io::IOContext{Base.Terminals.TerminalBuffer}, x::SubString{AnnotatedString{String}})
│        @ Base.AnnotatedDisplay strings/annotated_io.jl:357 [inlined]
│     [13] write(io::IOContext{Base.Terminals.TerminalBuffer}, s::SubString{AnnotatedString{String}})
│        @ Base.AnnotatedDisplay strings/annotated_io.jl:360 [inlined]
│     [14] write(io::IOContext{Base.Terminals.TerminalBuffer}, s::SubString{AnnotatedString{String}})
│        @ Base strings/annotated_io.jl:68 [inlined]
│     [15] refresh_multi_line(termbuf::Base.Terminals.TerminalBuffer, terminal::Base.Terminals.UnixTerminal, buf::IOBuffer, state::REPL.LineEdit.InputAreaState, prompt::Any; indent::Int64, region_active::Bool, show_cursor::Bool)
│        @ REPL.LineEdit ~/.julia/juliaup/julia-nightly/share/julia/stdlib/v1.14/REPL/src/LineEdit.jl:896
│     [16] refresh_multi_line(termbuf::Base.Terminals.TerminalBuffer, terminal::Base.Terminals.UnixTerminal, s::REPL.LineEdit.PromptState; beeping::Bool, show_cursor::Bool)
│        @ REPL.LineEdit ~/.julia/juliaup/julia-nightly/share/julia/stdlib/v1.14/REPL/src/LineEdit.jl:2515 [inlined]
│     [17] refresh_multi_line(termbuf::Base.Terminals.TerminalBuffer, terminal::Base.Terminals.UnixTerminal, s::Union{REPL.LineEdit.PrefixSearchState, REPL.LineEdit.PromptState})
│        @ REPL.LineEdit ~/.julia/juliaup/julia-nightly/share/julia/stdlib/v1.14/REPL/src/LineEdit.jl:2512
│     [18] refresh_multi_line(terminal::Base.Terminals.UnixTerminal, args::Any; kwargs...)
│        @ REPL.LineEdit ~/.julia/juliaup/julia-nightly/share/julia/stdlib/v1.14/REPL/src/LineEdit.jl:950
│     [19] refresh_multi_line(terminal::Base.Terminals.UnixTerminal, args::Any)
│        @ REPL.LineEdit ~/.julia/juliaup/julia-nightly/share/julia/stdlib/v1.14/REPL/src/LineEdit.jl:947
│     [20] refresh_multi_line(s::REPL.LineEdit.PromptState; kw...)
│        @ REPL.LineEdit ~/.julia/juliaup/julia-nightly/share/julia/stdlib/v1.14/REPL/src/LineEdit.jl:780
│     [21] refresh_multi_line(s::REPL.LineEdit.PromptState)
│        @ REPL.LineEdit ~/.julia/juliaup/julia-nightly/share/julia/stdlib/v1.14/REPL/src/LineEdit.jl:770
│     [22] refresh_multi_line(::REPL.LineEdit.MIState; kwargs...)
│        @ REPL.LineEdit ~/.julia/juliaup/julia-nightly/share/julia/stdlib/v1.14/REPL/src/LineEdit.jl:459
│     [23] refresh_line(s::REPL.LineEdit.MIState)
│        @ REPL.LineEdit ~/.julia/juliaup/julia-nightly/share/julia/stdlib/v1.14/REPL/src/LineEdit.jl:1899 [inlined]
│     [24] (::REPL.LineEdit.var"#59#60"{NTuple{9, Char}, Char, Char})(::REPL.LineEdit.MIState, ::Any, ::Vararg{Any})
│        @ REPL.LineEdit ~/.julia/juliaup/julia-nightly/share/julia/stdlib/v1.14/REPL/src/LineEdit.jl:2390
│     [25] (::REPL.LineEdit.var"#match_input##0#match_input##1"{REPL.LineEdit.var"#59#60"{NTuple{9, Char}, Char, Char}, String})(s::Any, p::Any)
│        @ REPL.LineEdit ~/.julia/juliaup/julia-nightly/share/julia/stdlib/v1.14/REPL/src/LineEdit.jl:2055
│     [26] macro expansion
│        @ ~/.julia/juliaup/julia-nightly/share/julia/stdlib/v1.14/REPL/src/LineEdit.jl:3252 [inlined]
│     [27] macro expansion
│        @ lock.jl:435 [inlined]
│     [28] (::REPL.LineEdit.var"#prompt!##2#prompt!##3"{Base.Terminals.TTYTerminal, REPL.LineEdit.ModalInterface, REPL.LineEdit.MIState, REPL.LineEdit.Prompt})()
│        @ REPL.LineEdit ~/.julia/juliaup/julia-nightly/share/julia/stdlib/v1.14/REPL/src/LineEdit.jl:3241
└ @ REPL.LineEdit ~/.julia/juliaup/julia-nightly/share/julia/stdlib/v1.14/REPL/src/LineEdit.jl:3254
```

**Comment by tecosaur** [comment]
I think this should be fixed now?
